azurerm/resource_arm_dns_mx_record.go

```diff
@@ -67,150 +68,140 @@ func resourceArmDnsMxRecord() *schema.Resource {
 		return err
 	}
-
-	flattenAndSetTags(d, &resp.Tags)
+	flattenAndSetTags(d, resp.Metadata)
 
 	return nil
 }
 
 func resourceArmDnsMxRecordDelete(d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*ArmClient)
-	rivieraClient := client.rivieraClient
-
-	deleteRequest := rivieraClient.NewRequestForURI(d.Id())
-	deleteRequest.Command = &dns.DeleteRecordSet{
-		RecordSetType: "MX",
-	}
+	dnsClient := meta.(*ArmClient).dnsClient
 
-	deleteResponse, err := deleteRequest.Execute()
+	id, err := parseAzureResourceID(d.Id())
 	if err != nil {
-		return fmt.Errorf("Error deleting DNS MX Record: %s", err)
+		return err
 	}
-	if !deleteResponse.IsSuccessful() {
-		return fmt.Errorf("Error deleting DNS MX Record: %s", deleteResponse.Error)
+
+	resGroup := id.ResourceGroup
+	name := id.Path["MX"]
+	zoneName := id.Path["dnszones"]
+
+	resp, error := dnsClient.Delete(resGroup, zoneName, name, dns.MX, "")
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("Error deleting DNS MX Record %s: %+v", name, error)
 	}
 
 	return nil
 }
 
-func expandAzureRmDnsMxRecord(d *schema.ResourceData) ([]dns.MXRecord, error) {
-	config := d.Get("record").(*schema.Set).List()
-	records := make([]dns.MXRecord, 0, len(config))
-
-	for _, pRaw := range config {
-		data := pRaw.(map[string]interface{})
-
-		mxrecord := dns.MXRecord{
-			Preference: data["preference"].(string),
-			Exchange:   data["exchange"].(string),
+// flatten creates an array of map where preference is a string to suit
+// the expectations of the ResourceData schema, so that this data can be
+// managed by Terradata state.
+func flattenAzureRmDnsMxRecords(records *[]dns.MxRecord) []map[string]interface{} {
+	results := make([]map[string]interface{}, 0, len(*records))
+
+	if records != nil {
+		for _, record := range *records {
+			preferenceI32 := *record.Preference
+			preference := strconv.Itoa(int(preferenceI32))
+			results = append(results, map[string]interface{}{
+				"preference": preference,
+				"exchange":   *record.Exchange,
+			})
 		}
-
-		records = append(records, mxrecord)
-
 	}
 
-	return records, nil
-
+	return results
 }
 
-func flattenAzureRmDnsMxRecord(records []dns.MXRecord) []map[string]interface{} {
-
-	result := make([]map[string]interface{}, 0, len(records))
-	for _, record := range records {
-		result = append(result, map[string]interface{}{
-			"preference": record.Preference,
-			"exchange":   record.Exchange,
-		})
+// expand creates an array of dns.MxRecord, that is, the array needed
+// by azure-sdk-for-go to manipulate azure resources, hence Preference
+// is an int32
+func expandAzureRmDnsMxRecords(d *schema.ResourceData) ([]dns.MxRecord, error) {
+	recordStrings := d.Get("record").(*schema.Set).List()
+	records := make([]dns.MxRecord, len(recordStrings))
+
+	for i, v := range recordStrings {
+		mxrecord := v.(map[string]interface{})
+		preference := mxrecord["preference"].(string)
+		i64, _ := strconv.ParseInt(preference, 10, 32)
```

Comment:
I think we should be able to update the schema and the set to be an `int` here - and then remove the `strconv` parse here?